### PR TITLE
Review/66

### PR DESCRIPTION
**Pull request by Husterknupp**
created this PR b/c #66 has to many single commits. Will do the review here condensed

PR does not have to be merged. It should ~~only~~ contain comments and potential updates. Given commits will be added, they should be **commited to branch `review/66` and `cherry-pick` ed later on** when the review is finished. 

Close this PR once the review is finished but don't merge to master.

Created like this:
```
git checkout 08cd917 # first before #66
git checkout -b review/66
git cherry-pick 9ab6847^..ac27a21
```

### CODE REVIEW

**Comment by Husterknupp** [comment]
@Bantolomeus assigning you as well so that you can accept code suggestions

**Comment by Bantolomeus** [comment]
It kinda sucks that not all changes are summarised under Files changed. It is still tedious to do the review. I am sorry for that =(

**Comment by Husterknupp** [comment]
@Bantolomeus I just created a second branch `review/66-to-compare-with` so that the diffs look good again

**Comment by Husterknupp** [comment]
@Bantolomeus I have commented on ProgressDTO.kt, ProgressRepository.kt and ProgressService.kt

see this for example:
screenshot 2018-11-11 at 00 12 34

don't you see the comments?

**Comment by Husterknupp** [comment]
@Bantolomeus again, I am now finished doing the review. Please respond to my comments and let's close the review + ticket

**Comment by Bantolomeus** [comment]
@Husterknupp could you please comment in the files changed tab and not every commit itself? It helps me to see all your comments and I do not need to search in every commit itself.

**Comment by Bantolomeus** [comment]
Since I see no comments I will close this pr.

**Comment by Bantolomeus** [comment]
`pagesSinceStart` will be deleted and not saved anymore but calculated on the fly. What would you suggest? challengePages? pagesDuringChallenge? pagesSinceChallengeBegin?

**Comment by Bantolomeus** [comment]
I really do not see your comments in this PR. Under the 15 changed files is nothing for me to see. I also did not get a mail as always.
Excuse me for closing the PR without answering to your last comment. I forget to answer. Was no offence.

**Comment by Bantolomeus** [comment]
I added travis ci as check to all PR that wants to merge into the master branch. Does not seem to bother this PR here.

**Comment by Husterknupp** [comment]
@Bantolomeus maybe there is no Travis here b/c this PR‘s base is not master 🤷‍♀️ could be..

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt [hunk 1]
```diff
@@ -100,48 +110,56 @@ class BookControllerIT {
...
+        assertEquals(readingStatusAfter, (readingStatusBefore + bookUpdateResponse.bookUpdates[0].pagesRead))
+        assertEquals(progressFileAfter.pagesEverRead, (progressFileBefore.pagesEverRead + 20))
         assertEquals(booksFile.books[0], bookDTOAfterUpdate)
-        assertEquals(bookResponse.bookUpdates[0].name, "Zap")
-        assertEquals(bookResponse.bookUpdates[0].date, today)
-        assertEquals(bookResponse.bookUpdates[0].pagesRead, 20)
-        assertEquals(bookResponse.bookUpdates[1].name, bookUpdates.bookUpdates[0].name)
-        assertEquals(bookResponse.bookUpdates[2].name, bookUpdates.bookUpdates[1].name)
-        assertEquals(bookResponse.bookUpdates[3].name, bookUpdates.bookUpdates[2].name)
-        assertEquals(bookResponse.bookUpdates[4].name, bookUpdates.bookUpdates[3].name)
```
[the remark this answers is not in the document]
Comment (Husterknupp) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: No

File: src/main/kotlin/com/bantolomeus/repository/ProgressRepository.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    fun saveProgress(progress: ProgressFileDTO): ProgressFileDTO {
+        objectMapper.writeValue(File(progressFile), progress)
+        return progress
+    }
+
+    fun getProgress(): ProgressFileDTO {
+        return try {
+            objectMapper.readValue(File(progressFile), ProgressFileDTO::class.java)
+        } catch (e: Exception) {
+            ProgressFileDTO()
```
[the remark this answers is not in the document]
Comment (Husterknupp) [reply src/main/kotlin/com/bantolomeus/repository/ProgressRepository.kt]: @Bantolomeus create issue to add error handling. Return optional (look this up).

File: src/main/kotlin/com/bantolomeus/service/ProgressService.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        progress.pagesEverRead = progress.pagesEverRead.plus(pages)
+        progress.pagesSinceStart = progress.pagesSinceStart.plus(pages)
+        return progressRepository.saveProgress(progress)
+    }
+
+    fun calculateReadingState(): Long {
+        val daysSinceStartNegated = -((Date().time - challengeRepository.getStartTime()) / DIVISOR_FOR_DAY)
+        val pagesPerDay = challengeRepository.getChallenge().pagesPerDay
+        val pagesSinceStart = progressRepository.getProgress().pagesSinceStart
+        return daysSinceStartNegated.times(pagesPerDay).plus(pagesSinceStart)
```
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/main/kotlin/com/bantolomeus/service/ProgressService.kt]: I will change it.

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Husterknupp) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: @Bantolomeus thanks for the open discussion, I like your answer

File: src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt [hunk 4]
```diff
@@ -1,3 +1,7 @@
 package com.bantolomeus.dto
 
-data class ProgressUpdateDTO(val date: String = "", val pagesRead: Long = 0)
+data class ProgressUpdateDTO(val date: String = "",
+                             val pagesRead: Long = 0)
+
+data class ProgressFileDTO(var pagesSinceStart: Long = 0,
```
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt]: 👍

**Comment by Bantolomeus** [comment]
@Husterknupp I find it kind of inconvenient with the cherry picking and so. I would like to commit the changes to master and add the commit to the comment, where that change belongs to. Is that ok or is that fucked up?

File: src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt]: I will change it

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt [hunk 5]
```diff
@@ -34,46 +34,56 @@ class BookControllerIT {
...
 
-        val challengeInit = challengeRepository.saveOrUpdateChallengeData(challengeDTO)
+        progressRepository.saveProgress(progressDTO)
+        challengeRepository.saveOrUpdateChallengeData(challengeDTO)
+
+        val readingStatusBefore = progressService.calculateReadingState()
         val bookResponse = bookController.createBook(bookDTO)
-        val challengeUpdated = challengeRepository.getChallenge()
+        val readingStatusAfter = progressService.calculateReadingState()
         val bookUpdatesUpdated = bookUpdateRepository.getBookUpdates()
```
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: I never read the message from the mail (I always go to the repo and look it up there), so most likely I would have never noticed the typo correction =) 
Yes, I will add this assertion.

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt [hunk 6]
```diff
@@ -100,48 +110,56 @@ class BookControllerIT {
...
+        assertEquals(bookUpdateResponse.bookUpdates[0].date, today)
+        assertEquals(bookUpdateResponse.bookUpdates[0].pagesRead, 20)
+        assertEquals(bookUpdateResponse.bookUpdates[1].name, bookUpdates.bookUpdates[0].name)
+        assertEquals(bookUpdateResponse.bookUpdates[2].name, bookUpdates.bookUpdates[1].name)
+        assertEquals(bookUpdateResponse.bookUpdates[3].name, bookUpdates.bookUpdates[2].name)
+        assertEquals(bookUpdateResponse.bookUpdates[4].name, bookUpdates.bookUpdates[3].name)
+        assertEquals(bookUpdateResponse.bookUpdates[5].name, bookUpdates.bookUpdates[4].name)
     }
 
     @Before
```
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: Do you mean that the `@Before` should come before the test cases? If yes I will move the `@Before` method to the top.

[the remark this answers is not in the document]
Comment (Husterknupp) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: yes this is what I mean

**Comment by Husterknupp** [comment]
@Bantolomeus just commit to master. This is fucked up anyway. Next time we do a proper branch/ PR also for backend changes,ok?

**Comment by Bantolomeus** [comment]
Agreed

File: src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/main/kotlin/com/bantolomeus/dto/ProgressDTO.kt]: changed it in https://github.com/Bantolomeus/bookmaster4000mkii/commit/3d080adc4d40e36bd2e75bd6fe087e82f2cffd02 
please resolve conversation if it is solved

File: src/main/kotlin/com/bantolomeus/service/ProgressService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/main/kotlin/com/bantolomeus/service/ProgressService.kt]: Changed it in https://github.com/Bantolomeus/bookmaster4000mkii/commit/3d080adc4d40e36bd2e75bd6fe087e82f2cffd02
please resolve conversation if it is resolved

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: Changed it in https://github.com/Bantolomeus/bookmaster4000mkii/commit/3d080adc4d40e36bd2e75bd6fe087e82f2cffd02
please resolve conversation if it is resolved

File: src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bantolomeus) [reply src/test/kotlin/com/bantolomeus/controller/BookControllerIT.kt]: Changed it in https://github.com/Bantolomeus/bookmaster4000mkii/commit/3d080adc4d40e36bd2e75bd6fe087e82f2cffd02
please resolve conversation if it is resolved

**Comment by Bantolomeus** [comment]
I changed everything as far as I was willing to. Please close the pr after everything is fine. If not comment again.

**Comment by Husterknupp** [comment]
@Bantolomeus just in case you missed it:

https://github.com/Bantolomeus/bookmaster4000mkii/pull/87#commitcomment-31244908

**Pull request closed without merging**